### DISCUSSION TITLE
send array of objects in supabase rpc parameters

### DISCUSSION THREAD

**Discussion by zakaria-chahboun** [supabase/supabase#3030] [Questions] [answered]
Hi everyone i have these tables,
* 1 question has many answers
![Screenshot from 2021-09-01 11-58-28](https://user-images.githubusercontent.com/18173275/131660182-ca1ec318-6b25-4d60-ac02-9f011231c08a.png)

i don't want to use `supabase.from(...).insert(...)` for each table! cuze i have to use it inside a transaction, i know that i have to use postgres functions and i calling it in my client with `supabase.rpc(...,{...})` but!

how i can send the `many answers` in function parameters?
and how i can insert these `many answers` inside the `answer` table! (i need a loop?)

Thanks.

**Comment by zakaria-chahboun** [comment, marked as answer]
I did that:
I create an answer type in SQL editor in dashboard: 
```sql
create TYPE answer_type AS (answer_title varchar, is_correct boolean);
```
Then i create the function `create_question` with all parameters i need, but in this example i gonna type just one parameter: answers *(an array of objects)*
```sql
create or replace function create_question(answers answer_type[])
returns varchar
language plpgsql
as $$
declare
answer_row answer_type;
begin

  FOREACH answer_row IN ARRAY answers
    LOOP
     -- Logic
     -- answer_row.answer_title
     -- answer_row.is_correct
    END LOOP;

  return "good";
end;$$
```
In client side, I send the answers parameter to my function like that:
```js
    const { data, error } = await supabase.rpc("create_question", {
      answers: [{answer_title:"answer one",is_correct:true},{answer_title:"answer 2",is_correct:false}]
    });
```

Is that the correct way?

**Reply by soedirgo** [reply]
Try if this works:
```js
    const { data, error } = await supabase.rpc("create_question", {
      answers: ["(answer one,true)","(answer 2,false)"]
    });
```
Either way creating a composite type is how I'd do it too.

**Reply by zakaria-chahboun** [reply]
My function is works fine, Just I'm asking if this is the only way! ☺️

**Reply by soedirgo** [reply]
I guess another way to do it is to take an array of `jsonb` instead, so you'd take `answers jsonb` and use [`jsonb_to_recordset`](https://www.postgresql.org/docs/current/functions-json.html#id-1.5.8.22.5.11.2.2.12.1.2.1) or similar to process the input. This way you don't need to create a user-defined type.

**Reply by zakaria-chahboun** [reply]
i already tried it, it depends also on a user-defined type.
```sql
select json_populate_recordset(null::answer_type, '[{"answer_title":"blablabla","is_correct":true},{"answer_title":"blablabla2","is_correct":false}]')
```

**Reply by soedirgo** [reply]
You don't have to deserialize the JSON into a type, just use the values directly:
```sql
insert into answer(answer_title, is_correct, question_id)
select answer_title, is_correct, 42
from jsonb_to_recordset('[{"answer_title":"a", "is_correct": true}, {"answer_title":"b", "is_correct": false}]')
```

**Reply by zakaria-chahboun** [reply]
I already tested it, it gives me this error message
```
a column definition list is required for functions returning "record"
```

**Reply by soedirgo** [reply]
Sorry, try this instead:
```sql
insert into answer(answer_title, is_correct, question_id)
select answer_title, is_correct, 42
from jsonb_to_recordset('[{"answer_title":"a", "is_correct": true}, {"answer_title":"b", "is_correct": false}]') x(answer_title text, is_correct boolean)
```

**Reply by zakaria-chahboun** [reply]
Yes that's cool!
```SQL
select
  answer_title,
  is_correct,
  42 as question_id
from
  jsonb_to_recordset(
    '[{"answer_title":"a", "is_correct": true}, {"answer_title":"b", "is_correct": false}]'
  ) foo(answer_title text, is_correct boolean)
```
It gives:
|answer_title|is_correct|question_id|
|--------------|-----------|-------------|
|a|true|42|
|b|false|42|

Cool, Thanks @soedirgo 
But i guess i will stick with the 1st way at this moment!

**Comment by zakaria-chahboun** [comment]
You can also use JSON without create a new custom type, like that [#discussioncomment-1270779](https://github.com/supabase/supabase/discussions/3030#discussioncomment-1270779)

Also if you have a nested object/array like this JSON:
> we have for example a set of questions, each question has a set of answers
```json
[
  {
    "question_text": "i'm question 1",
    "answers": [
      {
        "answer_text": "i'm answer 1",
        "is_correct": true
      },
      {
        "answer_text": "i'm answer 2",
        "is_correct": false
      }
    ]
  },
  {
    "question_text": "i'm question 2",
    "answers": [
      {
        "answer_text": "i'm answer 1",
        "is_correct": false
      },
      {
        "answer_text": "i'm answer 2",
        "is_correct": true
      }
    ]
  }
]
```
We can create a function to do that:
```sql
create or replace function create_question(my_json varchar)
returns json
language plpgsql
as $$
declare
question_row record; -- to store value of iterator (of each question) 
answer_row record; -- to store value of iterator (of each answer)
question_id int; -- to return id of new question 
begin
-- Do some logic here
  for question_row in select * from jsonb_to_recordset(my_json::jsonb) as x(question_text varchar, answers varchar)
  loop
    -- [Do want you want here] question_row.question_text, question_row.answers
    for answer_row in select * from jsonb_to_recordset(question_row.answers::jsonb) as x(answer_text varchar, is_correct boolean)
    loop
    -- [Do want you want here] answer_row.answer_text, answer_row.is_correct;
    end loop;
  end loop;

-- RETURN
return FORMAT('{"question_id":%s}', question_id);
-- In case of an error, fail:
  RAISE EXCEPTION 'create_question refuse'
    USING DETAIL = 'cannot run create_question properly',
          HINT = 'There something not ok.';
-- The above will cause a rollback
end;$$
```

Now let's test our function with RCP:
```js
let { data, error } = await supabase
  .rpc('create_question', {
    my_json: [YOUR_JSON_HERE]
  })

if (error) console.error(error)
else console.log(data)
```
